DYN-1874: I want to download missing graph package dependencies easily

scottmitchell opened the pull request:
### Purpose

This PR addresses JIRA task [DYN-1874](https://jira.autodesk.com/browse/DYN-1874). It adds an API to the `PackageManagerClientViewModel` to initiate download and install of a package. It also adds the ability for the Dependency Viewer to download missing packages.

![DepViewer2](https://user-images.githubusercontent.com/11542519/62078988-c033eb80-b21b-11e9-8ad5-7d7b013bc5af.PNG)


![Screen Recording 2019-07-26 at 2 59 20 PM_4](https://user-images.githubusercontent.com/11542519/61976387-ad21e100-afb9-11e9-90c0-f8a40f7b60fe.gif)


### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] ~~Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.~~ **This feature is in Preview Mode**

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

mjkkirschner commented on src/WorkspaceDependencyViewExtension/WorkspaceDependencyView.xaml.cs:
```diff
@@ -18,6 +20,8 @@ public partial class WorkspaceDependencyView : UserControl
 
         private WorkspaceModel currentWorkspace;
 
+        private DynamoViewModel dynamoViewModel;
```
don't do this please! let's figure out what we need to add to the extensions API to accomplish this without passing around the entire viewModel.

QilongTang replied:
@mjkkirschner We need to access `PackageManagerClientViewModel` to download package here, what do you recommend? Seems we could just expose the download package part to be an API, there are a few examples there, @scottmitchell

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:
```diff
@@ -441,6 +443,212 @@ public List<PackageManagerSearchElement> ListAll()
+            return String.Join(", ", pkgs.Select(x => x.Name + " " + x.VersionName));
+        }
+
+        internal void ExecutePackage(string name, PackageVersion version, string downloadPath)
```
what does this name mean?

QilongTang replied:
@scottmitchell You probably mean `DownloadPackage` here?

scottmitchell replied:
Most of this logic was previously in a method called `PackageOnExecuted` on `PackageManagerSearch` before I moved it to the `PackageManagerClient` in order for it to be accessed from other extensions. The name was a vestige of that. I've changed it to `ExecutePackageDownload`.

QilongTang commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:
```diff
@@ -441,6 +443,212 @@ public List<PackageManagerSearchElement> ListAll()
+            if (version == null)
+            {
+                // TODO: move message to resources
+                var result = MessageBox.Show(string.Format("Version {0} of {1} could not be found.", packageInfo.Version.ToString(), packageInfo.Name),
```
Do you still intend to finish the TODO in this PR?

scottmitchell replied:
Yes, thanks for pointing this out. I have moved them now

QilongTang commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:
```diff
@@ -441,6 +443,212 @@ public List<PackageManagerSearchElement> ListAll()
+                var headers = version.full_dependency_ids.Select(dep => dep._id).Select((id) =>
+                {
+                    PackageHeader pkgHeader;
+                    var res = pmExt.PackageManagerClient.DownloadPackageHeader(id, out pkgHeader);
```
Since you are already in the PackageManagerClientViewModel, I think you no longer need to access all the way from extension, `Model` here should be client right?

QilongTang commented:
A few comments, but looks pretty good so far and thank you for adding row details function in an early stage. Since there are lots of new style changes, have you looked at reusing color defined from Dynamo style sheets instead of using hex codes?

mjkkirschner commented on src/DynamoCoreWpf/Extensions/ViewLoadedParams.cs:
```diff
@@ -49,6 +49,11 @@ public IRenderPackageFactory RenderPackageFactory
             get { return dynamoViewModel.RenderPackageFactoryViewModel.Factory; }
         }
 
+        public PackageManagerClientViewModel PackageManagerClientViewModel
```
I don't think we should do this either 😉  - sorry for not being able to respond earlier - 
If there's no better way, than okay, but lets think more generally, are we going to do this for every view Model in Dynamo? It exposes too much of the application IMO and doesn't help people find what they need.

This seems like a very specific large surface to expose, and it exposes quite a lot of APIs ... 
What about adding just the apis you need (publish, download etc?) or event an event to request those?

mjkkirschner replied:
another way to think about it is to follow the example set out by the rest of the extension API - everything is exposed is an interface that describes a set of small apis that we consider required or useful, then you expose the object, but only as that interface, that way you control what apis we support.

mjkkirschner replied:
look at `IworkspaceModel` as an example, the extensions interfaces don't expose `WorkspaceModel` but give users an `IworkspaceModel` - they can cast this if they choose, but API users are making a decision to trade some safety for flexibility, and that is more clear.

scottmitchell replied:
This makes sense. All we need to surface right now is "DownloadPackage". Let's discuss briefly tomorrow.

mjkkirschner commented on src/WorkspaceDependencyViewExtension/WorkspaceDependencyView.xaml.cs:
```diff
@@ -170,137 +118,81 @@ internal void DependencyRegen(WorkspaceModel ws)
-    /// Class defining data for each column
-    /// </summary>
-    public class ColumnData
+    public class NodeLibraryDependencyRow
```
did something weird happen to this file? The diff is pretty hard to read, it looks like a bunch of public classes were deleted? Or am I mis-reading the diff?

mjkkirschner replied:
(you can't delete the public classes if they get released with another version - however short lived that version is) it's safer to just obsolete them as we don't know what binary compatibility issues this will cause without extensive testing, and it's just not worth it IMO.

QilongTang replied:
@mjkkirschner Do you remember the last case this was an issue? This class is only used for data binding to the UI so we can't make it internal. If any class went out with a beta release that we can no longer rename, that sounds like too strict..

mjkkirschner replied:
I'm not talking about beta releases.

If it's in 2.3 then we should consider it stable, unless @scottmitchell wants to refactor 2.3 before release to not include this class and then redo testing there.

QilongTang replied:
Sure, I guess I am wondering if the API restriction need to be applied on extension level. I understand once it is in RC branch we should make it stable but this dll is not in our public nuget so I do not expect developers to consume it nor us maintaining the stability.. My 2 cents..

mjkkirschner replied:
@QilongTang - that's a great great point! I think it's acceptable to make this entire .dll **not** part of the API.

👍 

I think we'll need to consider what consequences this has technically and also for build processes though.

QilongTang replied:
@mjkkirschner I don't think we have it specified according quickly glancing at https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions or https://github.com/DynamoDS/Dynamo/wiki/API-Changes. And seems we also missed 2.x versions' API changes...

mjkkirschner replied:
regarding the API Changes doc... I find it kind of useless work - the API diff job now computes this for us - or something like
https://www.fuget.org/packages/DynamoVisualProgramming.Core/2.2.1.5204/lib/net47/diff/2.1.0.7465/

is SOO much better than that page - it's only missing the explanations/readme portion.

scottmitchell commented:
@mjkkirschner @QilongTang I think I've addressed all the comments in this PR. In terms of the API / semantic versioning issue for the WorkspaceDependencyViewExtension--I think this will be addressed by release notes and by a PR to 2.3 marking the view extension (and/or its classes) as obsolete/beta.

scottmitchell commented on src/WorkspaceDependencyViewExtension/WorkspaceDependencyView.xaml.cs:
```diff
@@ -170,137 +118,112 @@ internal void DependencyRegen(WorkspaceModel ws)
+                switch (DependencyInfo.State)
+                {
+                    case PackageDependencyState.Loaded:
+                        message = string.Format("{0} {1} is installed locally.", 
```
@mjkkirschner @QilongTang While this view extension is in "Preview", can we leave user facing strings in the code rather than move them to resources? I imagine these messages will be changes as we add other features such as "Keep Local"

mjkkirschner replied:
thats okay with me as long as tasks are filed and TODOs added referencing them.

scottmitchell replied:
Ok, sounds like less work to just go ahead and move them to resources, so I'll do that

scottmitchell replied:
Moved to resx

mjkkirschner commented on src/DynamoCore/Graph/Workspaces/PackageDependencyInfo.cs:
```diff
@@ -85,7 +101,7 @@ public override string ToString()
     }
 
-    internal enum ReferenceType
+    public enum ReferenceType
```
does this need to be public?

scottmitchell replied:
Nope. Good point. Fixing this.

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/PackageManager/IPackageDownloader.cs:
```diff
@@ -0,0 +1,22 @@
+        /// Initiates download and install of a package
+        /// </summary>
+        /// <param name="package"></param>
+        /// <param name="downloadPath"></param>
```
can you add comments here? If there is nothing to say about these params leave them out of the /// comments.

mjkkirschner commented on src/WorkspaceDependencyViewExtension/WorkspaceDependencyView.xaml.cs:
```diff
@@ -170,137 +118,112 @@ internal void DependencyRegen(WorkspaceModel ws)
+                        break;
+
+                    case PackageDependencyState.Missing:
+                        iconPath = "NodeLibraryDependency_Missing.png";
```
would it be possible to move these to a resx file? I guess we can do it as as followup - raw resources like this are difficult to manage and make patch installer issues much worse.

mjkkirschner replied:
for example look at:
https://github.com/DynamoDS/Dynamo/blob/master/src/Libraries/GeometryUIWpf/ProtoGeometryImages.resx

scottmitchell replied:
@mjkkirschner I spent some time trying to figure this out last night, but couldn't quite get it to work. If it's alright with you, I'd like to deal with this in a follow up UI pass.

scottmitchell replied:
@mjkkirschner Finally got it. Turns out we have an existing utility for converting bitmaps to ImageSources: [https://github.com/DynamoDS/Dynamo/blob/ec10f936824152e7dd7d6d019efdcda0d78a5264/src/DynamoCoreWpf/Utilities/ResourceUtilities.cs](https://github.com/DynamoDS/Dynamo/blob/ec10f936824152e7dd7d6d019efdcda0d78a5264/src/DynamoCoreWpf/Utilities/ResourceUtilities.cs)

mjkkirschner commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:
```diff
@@ -165,7 +167,7 @@ private void ShowTermsOfUseForPublishing()
     ///     the Package Manager
     /// </summary>
-    public class PackageManagerClientViewModel : NotificationObject
+    public class PackageManagerClientViewModel : NotificationObject, IPackageInstaller
```
I don't think it is - A new method has been added, and this object now implements a new interface with that method. What API does it break?

aparajit-pratap replied:
If there are derived classes they would need to implement the new interface.

aparajit-pratap replied:
I guess it might be okay. I take it back. API breaking rules always confound me.

scottmitchell commented on src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerClientViewModel.cs:
```diff
@@ -441,6 +443,218 @@ public List<PackageManagerSearchElement> ListAll()
+            return String.Join(", ", pkgs.Select(x => x.Name + " " + x.VersionName));
+        }
+
+        internal void ExecutePackageDownload(string name, PackageVersion version, string downloadPath)
```
I agree this would be better, but I'd like to wrap up this task so I'm not going to do this right now.

scottmitchell commented:
@aparajit-pratap @mjkkirschner Thank you for the comments. I believe I have addressed them all--the one thing I did not fix here that will need a follow up is moving the package state icons to a resx file

scottmitchell commented:
@mjkkirschner I have moved the dependency state icons to a resx, and also added a "Package Manager Terms of Use" check to the new package download API (thanks for pointing this out @kronz). This PR should be good to go now.

QilongTang commented:
The TOU check and image changes looks good. Thank you for removing the static images so they will not pollute binary compatibility.  Merging for testing in 2.5.
